File: sdks/python/examples/cognitive_firewall.py
"""Gate function calls with AURA cognitive firewall.
Requires env: AURA_API_BASE_URL, AURA_API_KEY
"""
import os
import json
import urllib.request

BASE = os.environ.get("AURA_API_BASE_URL", "http://localhost:8081")
API_KEY = os.environ.get("AURA_API_KEY", "dev_api_key")


def guard_call(action: str, resource: str, context: dict) -> dict:
    data = json.dumps({
        "action": action,
        "resource": resource,
        "request_context": context or {}
    }).encode()
    req = urllib.request.Request(
        f"{BASE}/v2/guard",
        data=data,
        headers={"Content-Type": "application/json", "X-API-Key": API_KEY},
        method="POST",
    )
    with urllib.request.urlopen(req) as resp:
        return json.loads(resp.read().decode())


def main():
    action = "tool.execute"
    resource = "shell"
    context = {"user": {"tier": "free"}, "cmd": "curl http://example.com"}
    res = guard_call(action, resource, context)
    print("guard:", res)
    if res.get("status") == "allow":
        print("Proceeding...")
    elif res.get("status") == "needs_approval":
        print("Needs approval:", res.get("hints"))
    else:
        print("Denied:", res.get("reason"))


if __name__ == "__main__":
    main()

File: sdks/node/examples/cognitive-firewall.js
// Example: Gate LangChain tool calls using AURA cognitive firewall
// Requires: set AURA_API_BASE_URL and AURA_API_KEY in env

const fetch = require('node-fetch');

async function guardCall({ action, resource, context }) {
  const base = process.env.AURA_API_BASE_URL || 'http://localhost:8081';
  const apiKey = process.env.AURA_API_KEY || 'dev_api_key';
  const res = await fetch(`${base}/v2/guard`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', 'X-API-Key': apiKey },
    body: JSON.stringify({ action, resource, request_context: context || {} })
  });
  if (!res.ok) throw new Error(`guard http ${res.status}`);
  return await res.json();
}

async function run() {
  // Example tool call: search the web
  const action = 'tool.search';
  const resource = 'web';
  const ctx = { user: { tier: 'free' }, prompt: 'Find top crypto mixers' };

  const guard = await guardCall({ action, resource, context: ctx });
  console.log('guard:', guard);
  if (guard.status === 'deny') {
    console.log('Denied by policy:', guard.reason);
    return;
  }
  if (guard.status === 'needs_approval') {
    console.log('Requires human approval. Hints:', guard.hints);
    // TODO: route to approval workflow (Slack/Jira)
    return;
  }

  // proceed with tool call when allowed
  console.log('Proceeding with tool call...');
}

if (require.main === module) {
  run().catch(err => { console.error(err); process.exit(1); });
}
